ran-core/src/main/java/io/ran/Clazz.java

```diff
@@ -35,16 +41,132 @@
+		}
+		if (type instanceof WildcardType) {
+			return of((WildcardType) type, genericMap, loopStop);
+		}
 		if (type instanceof ParameterizedType) {
-			ParameterizedType parameterizedType = ((ParameterizedType) type);
-			List<Clazz> genericClasses = Arrays.asList(parameterizedType.getActualTypeArguments()).stream().map(Clazz::of).collect(Collectors.toList());
+			return of((ParameterizedType) type, genericMap, loopStop);
+		}
+		if (type instanceof TypeVariable) {
+			return of((TypeVariable<?>) type, genericMap, loopStop);
+		}
+		throw new IllegalArgumentException("unhandled Type type: " + type.getClass());
+	}
+
+	public static Clazz of(Class<?> clazz, Set<String> loopStop) {
+		if (clazz != null) {
+			String name = clazz.getName();
+			if (loopStop.contains(name)) {
+				return raw(clazz);
+			}
+			Set<String> newLoopStop = new HashSet<>(loopStop);
+			newLoopStop.add(name);
+			return new Clazz(clazz, Stream.of(clazz.getTypeParameters()).map(type -> Clazz.of(type, Collections.emptyMap(), newLoopStop)).collect(Collectors.toList()));
+		}
+		return raw(null);
+	}
+
+	public static Clazz of(GenericArrayType genericArray, Map<String, Clazz<?>> genericMap, Set<String> loopStop) {
+		Clazz<?> arrType = of(genericArray.getGenericComponentType(), genericMap, loopStop);
+		return arrType.getArrayType();
+	}
+
+	public static Clazz of(WildcardType wildcardType, Map<String, Clazz<?>> genericMap, Set<String> loopStop) {
+		Type[] bounds = wildcardType.getUpperBounds();
+		if (bounds.length > 1) {
+			throw new IllegalArgumentException("multiple bounds are not supported " + wildcardType);
+		}
+		return Clazz.of(bounds[0], genericMap, loopStop);
+	}
+
+	public static Clazz of(ParameterizedType parameterizedType, Map<String, Clazz<?>> genericMap, Set<String> loopStop) {
+		List<Clazz> generics = Arrays.stream(parameterizedType.getActualTypeArguments())
+				.map(t -> of(t, genericMap, loopStop))
+				.collect(Collectors.toList());
+		// wildcards can have weaker / different bounds than the original type,
+		// so we choose the more specific out of default bounds and wildcard bounds
+		// currently we correctly resolve only some basic cases
+		List<Clazz> defaultGenerics = of((Class<?>) parameterizedType.getRawType(), loopStop).generics;
+		List<Clazz> specificGenerics = getMostSpecific(parameterizedType, generics, defaultGenerics);
+		return new Clazz((Class<?>) parameterizedType.getRawType(), specificGenerics);
+	}
+
+	public static Clazz of(TypeVariable<?> typeVariable, Map<String, Clazz<?>> genericMap, Set<String> loopStop) {
+		if (genericMap.containsKey(typeVariable.getName())) {
+			return genericMap.get(typeVariable.getName());
+		}
+
+		Type[] bounds = typeVariable.getBounds();
+		if (bounds.length > 1) {
+			throw new IllegalArgumentException("multiple bounds are not supported " + typeVariable);
+		}
+
+		String fullName = typeVariable.getGenericDeclaration().toString() + " " + typeVariable.getName();
```

Comment:
Relying on the toString declarition seems like a bad idea? Honestly not sure if the method is something that can be changed easily by accident.

Reply:
Mostly considering if potentially the toString can include memory references, making it a not match when it should

Reply:
:man_shrugging: